File: solutions/3702-longest-subsequence-with-non-zero-bitwise-xor/bit-manipulation.cpp
// time: O(n)
// space: O(1)

#include <vector>
using namespace std;

class Solution {
  public:
    int longestSubsequence(vector<int> nums) {
        const int n = nums.size();
        int x = 0;
        int zeroes = 0;

        for (int i = 0; i < n; i++) {
            const int num = nums[i];
            if (num == 0)
                zeroes++;
            else
                x ^= num;
        }

        if (x) return n;
        if (zeroes == n) return 0;
        return n - 1;
    }
};

File: solutions/3702-longest-subsequence-with-non-zero-bitwise-xor/bit-manipulation.js
// time: O(n)
// space: O(1)

/**
 * @param {number[]} nums
 * @return {number}
 */
const longestSubsequence = nums => {
    const n = nums.length;
    let result = n;
    let xor = 0;
    let zeroes = 0;

    for (let i = 0; i < n; i++) {
        const num = nums[i];
        if (num === 0) zeroes++;
        else xor ^= num;
    }

    if (xor) return result;
    if (zeroes === n) return 0;
    return result - 1;
};
